### PR TITLE
Normalize tests invocation

### PR DESCRIPTION
**Pull request by marusak**
I am just testing this. And with having opened PR that has these changes (when this lands, all PRs needs to rebase) it is easier to test.

Blocked on:

- [x] #12368
- [x] #12371

### CODE REVIEW

File: bots/tests-invoke [hunk 1]
```diff
@@ -183,7 +181,7 @@ class PullTask(object):
             # explicit request for html logs
             status["link"] = "log.html"
             status["extras"] = [ "https://raw.githubusercontent.com/cockpit-project/cockpit/master/bots/task/log.html" ]
-        elif self.context.startswith("cockpit/"):
+        elif self.test_project:
```
Comment (martinpitt) [inline bots/tests-invoke]: This will regress the log.html bits for cockpit further down below, right? At least we want test_project to be something like `cockpit-project/cockpit` for cockpit itself, or isn't that the case yet?

File: test/run [hunk 2]
```diff
@@ -0,0 +1,103 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
```
Comment (martinpitt) [inline test/run]: This isn't necessary for py3. (And can be removed from all other py3 files).

Please split out this commit into a separate PR. This by itself is harmless, we can test it locally and land it fast, and then start rebasing all other PRs onto it, so that the actual change of the test names then doesn't cause so much trouble.

File: test/run [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+
+# This file is part of Cockpit.
+#
+# Copyright (C) 2016 Red Hat, Inc.
```
Comment (martinpitt) [inline test/run]: bump year

File: test/run [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with Cockpit; If not, see <http://www.gnu.org/licenses/>.
+
+import os
+import subprocess
+import sys
+
+
+# FIXME: Learn image-prepare to read image from TEST_OS
```
Comment (martinpitt) [inline test/run]: Why is that necessary?

File: test/run [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+    sys.stderr.write("Preparing image: building and installing Cockpit ...\n")
+
+    try:
+        # Download all the additional images so that even older branches find them
+        subprocess.check_call([os.path.join("bots", "image-download"),
+                               "candlepin", "fedora-stock", "fedora-23-stock",
+                               "ipa", "openshift", "selenium"])
+
+    except subprocess.CalledProcessError:
+        return "Downloading of additional images failed"
```
Comment (martinpitt) [inline test/run]: Let's please drop this now. It's annoying enough locally, and hasn't been necessary with our current CI for ages.

File: test/run [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+
+    # COMPAT: Create a legacy tmp/run directory to prevent races during testing
+    os.makedirs(os.path.join("test", "tmp", "run"), exist_ok=True)
```
Comment (martinpitt) [inline test/run]: This can also go. bots/machine/machine_core/machine_virtual.py `_start_qemu()` creates it if necessary.

File: test/run [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+
+    # Now actually run the prepare tooling
+    cmd = [os.path.join("test", "image-prepare")]
+    if prefix in ["image", "container"]:
```
Comment (martinpitt) [inline test/run]: What is this "image" prefix? I believe this should be dropped.

File: test/run [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+        cmd += ["--containers"]
+    cmd += ["--verbose"]
+
+    # Flush our own output before invoking the command
+    sys.stdout.flush()
+    sys.stderr.flush()
+
+    try:
+        # Do the basic prepare
+        subprocess.check_call(cmd + [image])
```
Comment (martinpitt) [inline test/run]: Won't this crash if `$TEST_OS` is unset and thus `image == None`?

File: test/run [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+
+    except subprocess.CalledProcessError:
+        return "Preparation of testable image failed"
+    return ""
+
+
+def main():
+    scenario = os.environ.get("TEST_SCENARIO", "verify")
+    jobs = os.environ.get("TEST_JOBS", 1)
+    image = os.environ.get("TEST_OS")
```
Comment (martinpitt) [inline test/run]: This should use `DEFAULT_IMAGE`.

File: test/run [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+
+    (prefix, _, value) = scenario.partition("-")
+    ret = prepare(prefix, value, image)
+
+    # Figure out what to do next
+    if prefix == "verify":
+        cmd = ["timeout", "120m", os.path.join("test", "verify", "run-tests"),
```
Comment (martinpitt) [inline test/run]: As discussed yesterday, let's please drop the timeouts here. The bots have a global timeout in tests-invoke, and each test has its own timeout, too. It was rather confusing last time having to change them in three different places.

File: bots/tests-invoke (hunk 1 above)
Comment (marusak) [reply bots/tests-invoke]: Right, I also though that `TEST_PROJECT` would be set up to `cockpit-project/cockpit` but it isn't.
This is what `tests-scan` outputs for cockpit test:
```
PRIORITY=0001 bots/make-checkout --verbose --base=master --repo=cockpit-project/cockpit pull/12361/head 0985a3bbfff2a97fa7f36d1a9253280db9dd3202 && GITHUB_BASE=cockpit-project/cockpit TEST_NAME=pull-12361-20190719-132222 TEST_REVISION=0985a3bbfff2a97fa7f36d1a9253280db9dd3202 bots/tests-invoke --pull-number=12361  --rebase=master fedora-30/selenium-chrome pull/12361/head
```
and this is for external project:
```
PRIORITY=0005 bots/make-checkout --verbose --base=master --repo=cockpit-project/starter-kit --bots-ref=pull/12361/head '' && TEST_PROJECT=cockpit-project/starter-kit TEST_NAME=pull-12361-20190719-132222 TEST_REVISION=0985a3bbfff2a97fa7f36d1a9253280db9dd3202 bots/tests-invoke --pull-number=12361  --rebase=master --remote=test centos-7 ''
```
I admit that I took this shortcut and just misused the fact that this is not set up.

File: test/run (hunk 4 above)
Comment (marusak) [reply test/run]: Just thought that majority of bots can read  `TEST_OS` from env, than maybe could also `image-prepare`

File: test/run (hunk 5 above)
Comment (marusak) [reply test/run]: What do you mean? Downloading of those additional images?

File: test/run (hunk 7 above)
Comment (marusak) [reply test/run]: I have no idea. It was just there and it most likely predates me as I cannot think of anything starting with `image`. Just din't feel confident enough to drop it. Removing

File: test/run (hunk 10 above)
Comment (marusak) [reply test/run]: All of them, or just this 120m? some of them have shorter timeout, not that it makes that much of a difference.

File: bots/tests-invoke (hunk 1 above)
Comment (martinpitt) [reply bots/tests-invoke]: Right, I figured something like that -- too many "cockpit" special cases. While we are at it, could you please test for cockpit here, and somewhere at the top make it default to cockpit, with a FIXME?

File: test/run (hunk 4 above)
Comment (martinpitt) [reply test/run]: Right, I don't mind adding, that, but I don't think it's required for this PR, is it?

File: test/run (hunk 5 above)
Comment (martinpitt) [reply test/run]: Yes, the entire paragraph.

File: test/run (hunk 7 above)
Comment (martinpitt) [reply test/run]: Ack, let's clean it up. Thanks!

File: test/run (hunk 10 above)
Comment (martinpitt) [reply test/run]: Yes, all of them. test/run can become quite simple then (this is why I thought a shell script would suffice, as it's really not much more than a `case` switch and calling a few programs -- but py is fine!).

File: bots/image-refresh [hunk 11]
```diff
@@ -28,79 +28,79 @@ from task import github, REDHAT_STORE
...
-        "cockpit/fedora-30/live-iso@weldr/lorax",
-        "cockpit/fedora-30/qcow2@weldr/lorax",
-        "cockpit/fedora-30/chrome@weldr/cockpit-composer",
-        "cockpit/fedora-30/firefox@weldr/cockpit-composer",
-        "cockpit/fedora-30/edge@weldr/cockpit-composer",
+        "fedora-30",
+        "fedora-30/selenium-chrome",
+        "fedora-30/selenium-firefox",
+        "fedora-30/selenium-edge",
+        "fedora-30/avocado-fedora",
```
Comment (marusak) [inline bots/image-refresh]: This should be just `fedora-30/avocado-fedora`, right? the `-fedora` is pointless there.

Comment (martinpitt) [reply bots/image-refresh]: Right, let's make this just `/avocado` (I suppose that's what you actually meant).

File: bots/task/github.py [hunk 12]
```diff
@@ -55,6 +55,7 @@
 # it will publish a test task to the queue (used to trigger specific contexts)
 NOT_TESTED_DIRECT = "Not yet tested (direct trigger)"
 
+# TODO this is not gonna work, is it?
```
Comment (marusak) [inline bots/task/github.py]: Now we won't be able to use this `OUR_CONTEXTS`, since tests are going to start always with `TEST_OS`. @martinpitt mentioned something about this and that he talked about it with @Gundersanne. What was the agreed solution?

Comment (martinpitt) [reply bots/task/github.py]: I had something else in mind when I said that we shouldn't need a static "cockpit/" prefix. Indeed we'll have to rework this. `known_context()` is used in tests-trigger as a sanity check to prevent us trying to e. g. retry semaphore with this (which won't work). But the more important user is `statuses()`.

So either `known_context()` checks if the first part (`.split('/')[0]`) exists (as link) in bots/images/ (some caching should be involved), or we have to use a static cockpit/ prefix. I'm just a bit concerned about test names getting too long, with the extra /branch name that we want to tack at the end. We could shorten it to "c/"..

But WDYT about the "is this a known image" check? It's not that expensive and easy to do.

Comment (marusak) [reply bots/task/github.py]: > But WDYT about the "is this a known image" check? It's not that expensive and easy to do.

I agree that checking if the first part is valid image makes sense. Adding `cockpit` could make some names way too long and it is still just noise. The same goes with `c/`. I'll take a look at this and send some commit.

**Comment by martinpitt** [comment]
Both prerequisite PRs landed, unblocking.

File: bots/task/github.py [hunk 13]
```diff
@@ -73,9 +64,10 @@
 TEAM_CONTRIBUTORS = "Contributors"
 
 def known_context(context):
-    for prefix in OUR_CONTEXTS:
-        if context.startswith(prefix):
-            return True
+    for project in testmap.projects():
+        for branch_tests in testmap.tests_for_project(project).values():
+            if context in branch_tests:
+                return True
```
Comment (marusak) [inline bots/task/github.py]: One problem with this - this will mark `fedora-30/verify` as unknown context although it is equivalent to `fedora-30` which would be a known context. Is this a problem and should this just split context at `\` and then split also all contexts from the map and check it exists?

File: bots/task/testmap.py [hunk 14]
```diff
@@ -17,97 +17,98 @@
...
+            'rhel-8-0-distropkg', 'rhel-8-1',
         ],
-        'rhel-7.6': ['avocado/fedora', 'container/kubernetes', 'container/bastion',
-            'selenium/firefox', 'selenium/chrome', 'verify/rhel-7-6',
+        'rhel-7.6': ['fedora-30/avocado', 'fedora-30/container-kubernetes', 'fedora-30/container-bastion',
+            'fedora-30/selenium-firefox', 'fedora-30/selenium-chrome', 'rhel-7-6',
         ],
-        'rhel-7.7': ['avocado/fedora', 'container/kubernetes', 'container/bastion',
-            'selenium/firefox', 'selenium/chrome', 'verify/rhel-7-7',
+        'rhel-7.7': ['fedora-30/avocado-fedora', 'fedora-30/container-kubernetes',
```
Comment (marusak) [inline bots/task/testmap.py]: containers, avocado and selenium are always F30 no matter which branch, right?

**Comment by marusak** [comment]
Rebased and edited to work with new changes. I've posted two comments in the code that I am not certain about.

File: bots/task/github.py (hunk 13 above)
Comment (martinpitt) [reply bots/task/github.py]: It's a bit unfortunate, but I don't think it will be a real-life problem. In principle we have this problem today already with non-cockpit problems. I think it's fine to start with the convention that the "verify" scenario isn't spelled out.

File: bots/task/testmap.py (hunk 14 above)
Comment (martinpitt) [reply bots/task/testmap.py]: I'd rather not make this assumption. We regularly update the image for these, and then also (have to) update the stable branches to keep up. But the point of this is to be more flexible, and always specify the OS. What's the background of your question?
